packages/project-build/src/shared/graph-utils.ts

```diff
@@ -0,0 +1,93 @@
+import type { Instance } from "@webstudio-is/sdk";
+
+type InstanceId = Instance["id"];
+
+// Depth-First Search (DFS) algorithm to find cycles in a directed graph
+export const findCycles = (
+  graph: Iterable<Pick<Instance, "id" | "children">>
+): InstanceId[][] => {
+  const adjacencyList: Record<InstanceId, InstanceId[]> = {};
```

Comment:
Maybe use map for it?

Reply:
Not needed, we are just iterating over it